mlrun/serving/routers.py

```diff
@@ -157,3 +163,256 @@ def _handle_event(self, event):
+
+        The `VotingEnsemble` applies the following logic:
+        Incoming Event -> Router Preprocessing -> Send to models (or model) ->
+        Apply all model logic (Preprocessing -> Prediction -> Postprocessing) ->
+        Router Voting logic -> Router Postprocessing -> Response
+
+        This enables you to do the general preprocessing and postprocessing steps
+        once on the router level, with only model-specific adjustments at the
+        model level.
+    """
+
+    def __init__(
+        self,
+        context,
+        name,
+        routes=None,
+        protocol=None,
+        url_prefix=None,
+        health_prefix=None,
+        vote_type=None,
+        executor_type=None,
+        **kwargs,
+    ):
+        super().__init__(
+            context, name, routes, protocol, url_prefix, health_prefix, **kwargs
+        )
+        self.url_prefix = url_prefix or f"/{self.protocol}/models"
+        self.name = "VoteRouter"
+        self.vote_type = vote_type
+        self.vote_flag = True if self.vote_type is not None else False
+        self.executor_type = executor_type
+        self._model_logger = _ModelLogPusher(self, context)
+        self.version = kwargs.get("version", "v1")
+        self.log_router = True
+
+    def _resolve_route(self, body, urlpath):
+        """Resolves the appropriate model to send the event to.
+        Supports:
+        - <prefix>/<model>[/versions/<ver>]/operation
+        Sends the event to the specific <model>[/versions/<ver>]
+
+        - <prefix>/operation
+        Sends the event to all models
+
+        Args:
+            body (dict): event body
+            urlpath (string): url path
+
+        Raises:
+            ValueError: [description]
+
+        Returns:
+            [type]: [description]
+        """
+        subpath = None
+        model = ""
+        if urlpath and not urlpath == "/":
+            # process the url <prefix>/<model>[/versions/<ver>]/operation
+            subpath = ""
+            urlpath = urlpath[len(self.url_prefix) :].strip("/")
+            if not urlpath:
+                # Only self.url_prefix was given
+                return "", None, ""
+            segments = urlpath.split("/")
+            if len(segments) == 1:
+                # Are we looking at a router level operation?
+                if segments[0] in ["infer", "predict", "explain"]:
+                    # We are given an operation
+                    return self.name, None, segments[0]
+            model = segments[0]
+            if len(segments) > 2 and segments[1] == "versions":
+                model = model + ":" + segments[2]
+                segments = segments[2:]
+            if len(segments) > 1:
+                subpath = "/".join(segments[1:])
+
+        if isinstance(body, dict):
+            # accepting route information from body as well
+            # to support streaming protocols (e.g. Kafka).
+            model = model or self.name
```

Comment:
I see now why you gave `self.name` a constant value, cause you wanted to be able signal yourself on a model name that simply representing the router itself, I suggest to separate between the two, i.e.
let name be name, and add in the `__init__` something like:
```
# A model name to represent the router itself
self.router_mask_model_name = "VoteRouter"
```
and use that `self.router_mask_model_name` in your functions instead of the name

Reply:
Added documentation